probe-rs/src/probe/blackmagic/arm.rs

```diff
@@ -0,0 +1,815 @@
+    ///
+    /// [ARM SVD Debug Description]: https://open-cmsis-pack.github.io/Open-CMSIS-Pack-Spec/main/html/debug_description.html#debugPortStart
+    fn debug_port_start(&mut self, dp: DpAddress) -> Result<(), ArmError> {
+        // Clear all errors.
+        // CMSIS says this is only necessary to do inside the `if powered_down`, but
+        // without it here, nRF52840 faults in the next access.
+        let mut abort = Abort(0);
+        abort.set_dapabort(true);
+        abort.set_orunerrclr(true);
+        abort.set_wderrclr(true);
+        abort.set_stkerrclr(true);
+        abort.set_stkcmpclr(true);
+        self.write_dp_register(dp, abort)?;
+
+        self.write_dp_register(dp, Select(0))?;
+
+        let ctrl = self.read_dp_register::<Ctrl>(dp)?;
+
+        let powered_down = !(ctrl.csyspwrupack() && ctrl.cdbgpwrupack());
+
+        if powered_down {
+            let mut ctrl = Ctrl(0);
+            ctrl.set_cdbgpwrupreq(true);
+            ctrl.set_csyspwrupreq(true);
+            self.write_dp_register(dp, ctrl.clone())?;
+
+            let start = Instant::now();
+            loop {
+                let ctrl = self.read_dp_register::<Ctrl>(dp)?;
+                if ctrl.csyspwrupack() && ctrl.cdbgpwrupack() {
+                    break;
+                }
+                if start.elapsed() >= Duration::from_secs(1) {
+                    return Err(ArmError::Timeout);
+                }
+                std::thread::sleep(Duration::from_millis(10));
+            }
+
+            // TODO: Handle JTAG Specific part
+
+            // TODO: Only run the following code when the SWD protocol is used
+
+            // Init AP Transfer Mode, Transaction Counter, and Lane Mask (Normal Transfer Mode, Include all Byte Lanes)
+            let mut ctrl = Ctrl(0);
+            ctrl.set_cdbgpwrupreq(true);
+            ctrl.set_csyspwrupreq(true);
+            ctrl.set_mask_lane(0b1111);
+            self.write_dp_register(dp, ctrl)?;
+
+            let ctrl_reg: Ctrl = self.read_dp_register(dp)?;
+            if !(ctrl_reg.csyspwrupack() && ctrl_reg.cdbgpwrupack()) {
+                tracing::error!("debug power-up request failed");
+                return Err(DebugPortError::TargetPowerUpFailed.into());
+            }
+
+            // According to CMSIS docs, here's where we would clear errors
+            // in ABORT, but we do that above instead.
+        }
+        Ok(())
+    }
+}
+
+impl ArmProbeInterface for BlackMagicProbeArmDebug {
+    fn access_ports(
+        &mut self,
+        dp: DpAddress,
+    ) -> Result<BTreeSet<FullyQualifiedApAddress>, ArmError> {
+        if dp != DpAddress::Default {
+            return Err(ArmError::NotImplemented("multidrop not yet implemented"));
+        }
+
+        Ok(self.access_ports.clone())
+    }
+
+    fn close(self: Box<Self>) -> Probe {
+        Probe::from_attached_probe(self.probe)
+    }
+
+    fn current_debug_port(&self) -> DpAddress {
+        todo!()
```

Comment:
This will lead to questionable UX. Can we find a way to implement this straight up? Or change the interface to return a Result worst case?